modules/packages/DistributedBag.chpl

```diff
@@ -0,0 +1,1084 @@
+    if block == nil then {
+      tailBlock = new BagSegmentBlock(eltType, distributedBagInitialBlockSize);
+      headBlock = tailBlock;
+      block = tailBlock;
+    }
+
+    // Full? Create a new one double the previous size
+    if block.isFull {
+      block.next = new BagSegmentBlock(eltType, block.cap * 2);
+      tailBlock = block.next;
+      block = block.next;
+    }
+
+    block.push(elt);
+    nElems.add(1);
+  }
+
+  inline proc addElements(elts) {
+    for elt in elts do addElements(elt);
+  }
+}
+
+/*
+  We maintain a multiset 'bag' per node. Each bag keeps a handle to it's parent,
+  which is required for work stealing.
+*/
+pragma "no doc"
+class Bag {
+  type eltType;
+
+  // A handle to our parent 'distributed' bag, which is needed for work stealing.
+  var parentHandle : DistributedBag(eltType);
+
+  /*
+    Helps evenly distribute and balance placement of elements in a best-effort
+    round-robin approach. In the case where we have parallel enqueues or dequeues,
+    they are less likely overlap with each other. Furthermore, it increases our
+    chance to find our 'ideal' segment.
+  */
+  var startIdxEnq : atomic uint;
+  var startIdxDeq : atomic uint;
+
+  /*
+    If a task makes 2 complete passes (1 best-case, 1 average-case) and has not
+    found enough items, then it may attempt to balance the load distribution.
+    Furthermore, if a task is waiting on a load balance, it may piggyback on the
+    result.
+  */
+  var loadBalanceInProgress : atomic bool;
+  var loadBalanceResult : atomic bool;
+
+
+
+  var maxParallelSegmentSpace = {0 .. #here.maxTaskPar};
+  var segments : [maxParallelSegmentSpace] BagSegment(eltType);
+
+  inline proc nextStartIdxEnq {
+    return (startIdxEnq.fetchAdd(1) % here.maxTaskPar : uint) : int;
+  }
+
+  inline proc nextStartIdxDeq {
+    return (startIdxDeq.fetchAdd(1) % here.maxTaskPar : uint) : int;
+  }
+
+  proc Bag(type eltType, parentHandle) {
+    this.parentHandle = parentHandle;
+  }
+
+  proc add(elt : eltType) : bool {
+    var startIdx = nextStartIdxEnq : int;
+    var phase = ADD_BEST_CASE;
+
+    while true {
+      select phase {
+        /*
+          Pass 1: Best Case
+
+          Find a segment that is unlocked and attempt to acquire it. As we are adding
+          elements, we don't care how many elements there are, just that we find
+          some place to add ours.
+        */
+        when ADD_BEST_CASE {
+          for offset in 0 .. #here.maxTaskPar {
+            ref segment = segments[(startIdx + offset) % here.maxTaskPar];
+
+            // Attempt to acquire...
+            if segment.acquireWithStatus(STATUS_ADD) {
+              segment.addElements(elt);
+              segment.releaseStatus();
+              return true;
+            }
+          }
+
+          phase = ADD_AVERAGE_CASE;
+        }
+        /*
+          Pass 2: Average Case
+
+          Find any segment (locked or unlocked) and make an attempt to acquire it.

```

Comment:
The comment says "Find any segment" but the average case code only looks in 1 segment. Is that intentional?

Reply:
Yep, that's outdated documentation. Originally we checked the status of the segment before trying to add to it (back when I had work stealing keep the lock on the segment, it would just look for an adjacent segment which wasn't). I can make a OR correcting it if needed

Reply:
Yep, that's outdated documentation. Originally we checked the status of the segment before trying to add to it (back when I had work stealing keep the lock on the segment, it would just look for an adjacent segment which wasn't). I can make a OR correcting it if needed

Reply:
Btw, has there been a reported issue so far with the collections?

Reply:
I havn't seen one, I'd let you know. I think we're still waiting for somebody to battle-test them.

You can fix the comment if you like, no big deal.